### PR TITLE
(PUP-10476) PAL additions to support more use cases from bolt

### PR DESCRIPTION
**Pull request by McdonaldSeanp**
This commit makes two changes to PAL to support more features within puppet during a catalog compile:

1. When variables are passed to `.with_catalog_compiler` they are now deserialized if they are serialized pcore types
2. A new function was added to the compiler class within `.with_catalog_compiler` so that a process calling PAL functions can evaluate ast with node definitions.

### CODE REVIEW

**Comment by donoghuc** [comment]
Raised this draft to show how we might use it in bolt. https://github.com/puppetlabs/bolt/pull/1798

File: lib/puppet/pal/catalog_compiler.rb [hunk 1]
```diff
@@ -97,6 +97,15 @@ def evaluate_additions
...
     end
 
+    # Attempts to evaluate AST for node defnintions https://puppet.com/docs/puppet/latest/lang_node_definitions.html
+    # if there are any.
+    def evaluate_ast_node
+      # This is definitely calling an internal function in the compiler class, we could update
+      # Puppet::Parser::Compiler to make that a public function, but I was concerned about
+      # changing something that could potentially be seen as public puppet API.
+      #                - Sean P. McDonald 5/7/2020
+      internal_compiler.send(:evaluate_ast_node)
```
Comment (McdonaldSeanp) [inline lib/puppet/pal/catalog_compiler.rb:107]: This is something we can change to be an actual public API if we want, I was just concerned about making more things public in puppet.

File: lib/puppet/pal/pal_impl.rb [hunk 2]
```diff
@@ -440,6 +439,10 @@ def self.main(manifest, facts, variables, internal_compiler_class)
         # TRANSLATORS: Do not translate, symbolic name
         Puppet.override(overrides, "PAL::with_#{internal_compiler_class}_compiler") do
           compiler.compile do | compiler_yield |
+            # In case the varaibles passed to the compiler are PCore types defined in modules, they
+            # need to be deserialized and added from within the this scope, so that loaders are
+            # available during deserizlization.
+            add_variables(compiler.topscope, Puppet::Pops::Serialization::FromDataConverter.convert(pal_variables))
```
Comment (donoghuc) [inline lib/puppet/pal/pal_impl.rb:445]: Why cant we add give https://github.com/puppetlabs/puppet/blob/f29b9681e72271ca7a1dc52932d566eb217ace52/lib/puppet/pal/catalog_compiler.rb#L10 an add_variables method? Then we could just call that in bolt/puppetserver?

File: lib/puppet/pal/catalog_compiler.rb (hunk 1 above)
Comment (justinstoller) [reply lib/puppet/pal/catalog_compiler.rb:107]: We can make it Ruby `public` while documenting it as `# @api private`. There's some prior art to doing that in the rest of Puppet.

File: lib/puppet/pal/pal_impl.rb (hunk 2 above)
Comment (McdonaldSeanp) [reply lib/puppet/pal/pal_impl.rb:445]: We can, but it seems incorrect to me that PAL would provide parameters for setting variables and not have them work for anything other than core puppet datatypes.

If PAL supports providing variables as a parameter to `.with_catalog_compiler` it should support those variables if they are pcore types defined in modules too.

**Comment by justinstoller** [comment]
I'm 💯 on updating PAL if it's not working for Bolt. Though I'll leave it to @puppetlabs/bolt to review if this is correct for them.

**Comment by nicklewis** [comment]
👍 This is great.

**Pull request merged**